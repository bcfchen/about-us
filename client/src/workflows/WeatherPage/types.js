import PropTypes from "prop-types";
import DailyForecast from "../../models/DailyForecast";
import Location from "../../models/Location";
import HourlyForecast from "../../models/HourlyForecast";

export const propTypes = {
    fiveDayForecasts: PropTypes.arrayOf(PropTypes.instanceOf(DailyForecast)).isRequired,
    hourlyForecasts: PropTypes.arrayOf(PropTypes.instanceOf(HourlyForecast)).isRequired,
    currentLocation: PropTypes.instanceOf(Location),
    isLoading: PropTypes.bool.isRequired
};