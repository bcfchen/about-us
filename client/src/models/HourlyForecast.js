import moment from "moment";

class HourlyForecast {
    constructor(properties, location) {
        this.temperature = properties.Temperature.Value;
        this.text = properties.IconPhrase;
        this.dateTime = properties.DateTime;
        this.windSpeed = properties.Wind.Speed.Value + " km/h";
        this.humidity = properties.RelativeHumidity + " %";
        this.uvIndex = properties.UVIndex;
        this.location = location;
    }

    getHoursAndMinutes() {
        return moment(this.dateTime).format("HH:mm A");
    }

    getLocationString() {
        return this.location.getLocationString();
    }
}

export default HourlyForecast;