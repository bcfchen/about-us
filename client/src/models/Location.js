class Location {
    constructor(city, region, country) {
        this.city = city;
        this.region = region;
        this.country = country;
    }

    getLocationString() {
        let hasInfo = this.city && this.region;
        return hasInfo ? this.city + ", " + this.region : "";
    }

    static fromLocationObj(locationObj) {
        return new Location(locationObj.city, locationObj.region, locationObj.country);
    }

    static fromLocationString(locationStr) {
        let locationArr = locationStr.split(", ");
        let city = locationArr[0],
            region = locationArr[1],
            country = locationArr[2];
        return new Location(city, region, country);
    }

    static fromLocationArr(locationArr) {
        let city = locationArr[1],
            region = locationArr[2],
            country = locationArr[3];
        return new Location(city, region, country);
    }
}

export default Location;