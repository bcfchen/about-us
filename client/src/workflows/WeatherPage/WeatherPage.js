import React from "react";
import { connect } from "react-redux";
import * as weatherActions from "../../redux/actions/weatherActions";
import * as locationActions from "../../redux/actions/locationActions";
import { bindActionCreators } from "redux";
import WeatherForecastList from "../../components/WeatherForecastList/WeatherForecastList";
import CurrentWeatherIndicator from "../../components/CurrentWeatherIndicator/CurrentWeatherIndicator";
import WeatherDetails from "../../components/WeatherDetails/WeatherDetails";
import WeatherChart from "../../components/WeatherChart/WeatherChart";
import LocationSearch from "../../components/LocationSearch/LocationSearch";
import ProgressIndicator from "../../components/ProgressIndicator/ProgressIndicator";

class WeatherPage extends React.Component {
    constructor(props, context) {
        super(props.context);
        this.loadWeather = this.loadWeather.bind(this);
        this.handleRefresh = this.handleRefresh.bind(this);
        this.getSuggestedLocations = this.getSuggestedLocations.bind(this);
        props.locationActions.loadCurrentLocation();
        props.weatherActions.loadCurrentLocationHourlyForecasts();
        props.weatherActions.loadFiveDayForecasts();
    }

    loadWeather(selectedLocation) {
        if (selectedLocation) {
            this.props.weatherActions.loadFiveDayForecasts(selectedLocation.value);
            this.props.weatherActions.loadHourlyForecasts(selectedLocation.value);
            this.props.locationActions.updateCurrentLocation(selectedLocation);
        }
    }

    getSuggestedLocations(inputText) {
        if (inputText && inputText.length > 0) {
            this.props.locationActions.updateCurrentLocation(inputText);
            this.props.locationActions.getSuggestedLocations(inputText);
        }
    }

    handleRefresh() {
        this.setState({ isLoading: true });
        this.props.locationActions.loadCurrentLocation();
        this.props.weatherActions.loadCurrentLocationHourlyForecasts();
        this.props.weatherActions.loadFiveDayForecasts();
    }

    getBodyClassName(currentHourConditions) {
        if (!currentHourConditions) { return "weather-body"; }
        return currentHourConditions.isDaylight ? "weather-body-day" : "weather-body-night";
    }

    render() {
        let blurblur = this.props.isLoading ? "blur" : "";
        let bodyClassName = this.getBodyClassName(this.props.hourlyForecasts[0]);
        return (
            <div className={bodyClassName}>
                {this.props.isLoading && <ProgressIndicator />}
                <div className={blurblur}>
                    <LocationSearch currentLocation={this.props.currentLocation}
                        onLocationSelected={this.loadWeather} suggestedLocations={this.props.locations} />
                    <CurrentWeatherIndicator handleRefresh={this.handleRefresh} isLoading={this.props.isLoading} currentHourlyForecast={this.props.hourlyForecasts[0]} />
                    <WeatherDetails isLoading={this.props.isLoading} currentHourlyForecast={this.props.hourlyForecasts[0]} />
                    <WeatherChart isLoading={this.props.isLoading} hourlyForecasts={this.props.hourlyForecasts} />
                    <WeatherForecastList isLoading={this.props.isLoading} weatherForecasts={this.props.fiveDayForecasts} />
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, ownProps) => {
    return {
        fiveDayForecasts: state.weatherPage.fiveDayForecasts,
        hourlyForecasts: state.weatherPage.hourlyForecasts,
        locations: state.weatherPage.locations,
        currentLocation: state.weatherPage.currentLocation,
        isLoading: state.ajaxCallsInProgress > 0
    };
}

const mapDispatchToProps = dispatch => {
    return {
        weatherActions: bindActionCreators(weatherActions, dispatch),
        locationActions: bindActionCreators(locationActions, dispatch)
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(WeatherPage);