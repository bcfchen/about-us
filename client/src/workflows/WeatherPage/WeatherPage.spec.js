import React from "react";
import { mount, shallow } from "enzyme";
import { WeatherPage } from "./WeatherPage";
import Location from "../../models/Location";
import toast from "toastr";
import HourlyForecast from "../../models/HourlyForecast";
jest.mock("toastr");

describe("WeatherPage", () => {
    // setup mock properties
    let defaultProps = {
        weatherActions: { loadFiveDayForecasts: () => Promise.resolve(), loadHourlyForecasts: () => Promise.resolve(), loadCurrentLocationHourlyForecasts: () => Promise.resolve() },
        locationActions: { loadCurrentLocation: () => Promise.resolve(), updateCurrentLocation: () => Promise.resolve() },
        fiveDayForecasts: [],
        hourlyForecasts: [],
        currentLocation: new Location("city", "region", "country", "key"),
        isLoading: false
    },
        shallowComponent = null,
        mountedComponent = null;

    let shallowRender = (props = defaultProps) => { shallowComponent = shallow(<WeatherPage {...props} />); };

    describe("on load", () => {
        it("should render correctly", () => {
            shallowRender();
            expect(shallowComponent).toMatchSnapshot();
        });

        describe("is still loading", () => {
            it("should be blurred", () => {
                shallowRender({ ...defaultProps, isLoading: true });
                expect(shallowComponent.find(".blur").exists()).toEqual(true);
            });
        });

        describe("is not still loading", () => {
            it("should not be blurred", () => {
                shallowRender({ ...defaultProps, isLoading: false });
                expect(shallowComponent.find(".blur").exists()).toEqual(false);
            });
        });

        describe("loading background given location time", () => {
            describe("if location is daytime", () => {
                it("should load daytime background", () => {
                    let hourlyForecastProps = {
                        Temperature: { Value: 15 },
                        Wind: { Speed: { Value: 5 } },
                        IsDaylight: true
                    };

                    shallowRender({ ...defaultProps, hourlyForecasts: [new HourlyForecast(hourlyForecastProps)] });
                    expect(shallowComponent.find(".weather-body-day").exists()).toEqual(true);
                });
            });

            describe("if location is nighttime", () => {
                it("should load nighttime background", () => {
                    let hourlyForecastProps = {
                        Temperature: { Value: 15 },
                        Wind: { Speed: { Value: 5 } },
                        IsDaylight: false
                    };

                    shallowRender({ ...defaultProps, hourlyForecasts: [new HourlyForecast(hourlyForecastProps)] });
                    expect(shallowComponent.find(".weather-body-night").exists()).toEqual(true);
                });
            });

            describe("if location is undefined", () => {
                it("should load default background", () => {
                    shallowRender({ ...defaultProps, hourlyForecasts: [] });
                    expect(shallowComponent.find(".weather-body").exists()).toEqual(true);
                });
            });
        });
    });

    describe("on refresh", () => {
        let resolve = jest.fn(),
            reject = jest.fn();

        it("should trigger resolve on success", () => {
            shallowRender();
            shallowComponent.find("ReactPullToRefresh").prop("onRefresh")(resolve, reject).then(() => {
                expect(resolve).toHaveBeenCalled();
            });
        });

        it("should trigger reject on error", () => {
            let errorProps = { ...defaultProps };
            errorProps.weatherActions.loadFiveDayForecasts = () => Promise.reject("error");
            shallowRender(errorProps);
            shallowComponent.find("ReactPullToRefresh").prop("onRefresh")(resolve, reject).then(() => {
                expect(reject).toHaveBeenCalled();
            });
        });
    });

    describe("on location selected", () => {
        describe("if valid location is provided", () => {
            it("actions should be called to reload weather data", () => {
                // given
                let mockWeatherActions = {
                    loadFiveDayForecasts: jest.fn(),
                    loadHourlyForecasts: jest.fn(),
                    loadCurrentLocationHourlyForecasts: jest.fn()
                };
                let overrideProps = { ...defaultProps };
                overrideProps.weatherActions = mockWeatherActions;
                overrideProps.locationActions.updateCurrentLocation = jest.fn();

                // when
                shallowRender(overrideProps);
                shallowComponent.find("WithStyles(LocationSearch)")
                    .prop("onLocationSelected")({ value: "locationVal" });

                // then
                expect(mockWeatherActions.loadFiveDayForecasts).toHaveBeenCalled();
                expect(mockWeatherActions.loadCurrentLocationHourlyForecasts).toHaveBeenCalled();
            })

            it("should call toast error if unsuccessful", () => {
                // given
                let overrideProps = { ...defaultProps };
                overrideProps.locationActions.updateCurrentLocation = jest.fn(() => Promise.reject("bad"));
                toast.error = jest.fn();
                shallowRender(overrideProps);

                // when
                shallowComponent.find("WithStyles(LocationSearch)")
                    .prop("onLocationSelected")({ value: "locationVal" }).then(() => {
                        // then
                        expect(toast.error).toHaveBeenCalled();
                    });
            })
        });

        describe("if valid location is not provided", () => {
            it("actions should not be called to reload weather data", () => {
                // given
                let mockWeatherActions = {
                    loadFiveDayForecasts: jest.fn(),
                    loadHourlyForecasts: jest.fn(),
                    loadCurrentLocationHourlyForecasts: jest.fn()
                };
                let overrideProps = { ...defaultProps };
                overrideProps.weatherActions = mockWeatherActions;
                overrideProps.locationActions.updateCurrentLocation = jest.fn();
                shallowRender(overrideProps);

                // when
                shallowComponent.find("WithStyles(LocationSearch)")
                    .prop("onLocationSelected")(undefined);

                // then 
                expect(mockWeatherActions.loadFiveDayForecasts).toHaveBeenCalledTimes(1); // once called by constructor
                expect(mockWeatherActions.loadHourlyForecasts).toHaveBeenCalledTimes(0);
            })
        });
    });

}); 