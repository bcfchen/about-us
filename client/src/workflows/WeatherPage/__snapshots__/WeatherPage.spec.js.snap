// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherPage on load should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <WeatherPage
    currentLocation={
        Location {
            "city": "city",
            "country": "country",
            "key": "key",
            "region": "region",
          }
    }
    fiveDayForecasts={Array []}
    hourlyForecasts={Array []}
    isLoading={false}
    locationActions={
        Object {
            "loadCurrentLocation": [Function],
            "updateCurrentLocation": [Function],
          }
    }
    weatherActions={
        Object {
            "loadCurrentLocationHourlyForecasts": [Function],
            "loadFiveDayForecasts": [Function],
            "loadHourlyForecasts": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <ReactPullToRefresh
        onRefresh={[Function]}
>
        <div
                className=""
        >
                <WithStyles(LocationSearch)
                        currentLocation={
                                Location {
                                        "city": "city",
                                        "country": "country",
                                        "key": "key",
                                        "region": "region",
                                      }
                        }
                        onLocationSelected={[Function]}
                />
                <CurrentWeatherIndicator
                        currentHourlyForecast={undefined}
                        isLoading={false}
                />
                <WeatherDetails
                        currentHourlyForecast={undefined}
                        isLoading={false}
                />
                <WeatherChart
                        hourlyForecasts={Array []}
                        isLoading={false}
                />
                <WeatherForecastList
                        isLoading={false}
                        weatherForecasts={Array []}
                />
        </div>
</ReactPullToRefresh>,
      "className": "weather-body",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          false,
          <div
            className=""
>
            <WithStyles(LocationSearch)
                        currentLocation={
                                    Location {
                                                "city": "city",
                                                "country": "country",
                                                "key": "key",
                                                "region": "region",
                                              }
                        }
                        onLocationSelected={[Function]}
            />
            <CurrentWeatherIndicator
                        currentHourlyForecast={undefined}
                        isLoading={false}
            />
            <WeatherDetails
                        currentHourlyForecast={undefined}
                        isLoading={false}
            />
            <WeatherChart
                        hourlyForecasts={Array []}
                        isLoading={false}
            />
            <WeatherForecastList
                        isLoading={false}
                        weatherForecasts={Array []}
            />
</div>,
        ],
        "onRefresh": [Function],
      },
      "ref": null,
      "rendered": Array [
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <WithStyles(LocationSearch)
                currentLocation={
                                Location {
                                                "city": "city",
                                                "country": "country",
                                                "key": "key",
                                                "region": "region",
                                              }
                }
                onLocationSelected={[Function]}
/>,
              <CurrentWeatherIndicator
                currentHourlyForecast={undefined}
                isLoading={false}
/>,
              <WeatherDetails
                currentHourlyForecast={undefined}
                isLoading={false}
/>,
              <WeatherChart
                hourlyForecasts={Array []}
                isLoading={false}
/>,
              <WeatherForecastList
                isLoading={false}
                weatherForecasts={Array []}
/>,
            ],
            "className": "",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "currentLocation": Location {
                  "city": "city",
                  "country": "country",
                  "key": "key",
                  "region": "region",
                },
                "onLocationSelected": [Function],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "currentHourlyForecast": undefined,
                "isLoading": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "currentHourlyForecast": undefined,
                "isLoading": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "hourlyForecasts": Array [],
                "isLoading": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "isLoading": false,
                "weatherForecasts": Array [],
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <ReactPullToRefresh
          onRefresh={[Function]}
>
          <div
                    className=""
          >
                    <WithStyles(LocationSearch)
                              currentLocation={
                                        Location {
                                                  "city": "city",
                                                  "country": "country",
                                                  "key": "key",
                                                  "region": "region",
                                                }
                              }
                              onLocationSelected={[Function]}
                    />
                    <CurrentWeatherIndicator
                              currentHourlyForecast={undefined}
                              isLoading={false}
                    />
                    <WeatherDetails
                              currentHourlyForecast={undefined}
                              isLoading={false}
                    />
                    <WeatherChart
                              hourlyForecasts={Array []}
                              isLoading={false}
                    />
                    <WeatherForecastList
                              isLoading={false}
                              weatherForecasts={Array []}
                    />
          </div>
</ReactPullToRefresh>,
        "className": "weather-body",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            false,
            <div
              className=""
>
              <WithStyles(LocationSearch)
                            currentLocation={
                                          Location {
                                                        "city": "city",
                                                        "country": "country",
                                                        "key": "key",
                                                        "region": "region",
                                                      }
                            }
                            onLocationSelected={[Function]}
              />
              <CurrentWeatherIndicator
                            currentHourlyForecast={undefined}
                            isLoading={false}
              />
              <WeatherDetails
                            currentHourlyForecast={undefined}
                            isLoading={false}
              />
              <WeatherChart
                            hourlyForecasts={Array []}
                            isLoading={false}
              />
              <WeatherForecastList
                            isLoading={false}
                            weatherForecasts={Array []}
              />
</div>,
          ],
          "onRefresh": [Function],
        },
        "ref": null,
        "rendered": Array [
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <WithStyles(LocationSearch)
                  currentLocation={
                                    Location {
                                                      "city": "city",
                                                      "country": "country",
                                                      "key": "key",
                                                      "region": "region",
                                                    }
                  }
                  onLocationSelected={[Function]}
/>,
                <CurrentWeatherIndicator
                  currentHourlyForecast={undefined}
                  isLoading={false}
/>,
                <WeatherDetails
                  currentHourlyForecast={undefined}
                  isLoading={false}
/>,
                <WeatherChart
                  hourlyForecasts={Array []}
                  isLoading={false}
/>,
                <WeatherForecastList
                  isLoading={false}
                  weatherForecasts={Array []}
/>,
              ],
              "className": "",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "currentLocation": Location {
                    "city": "city",
                    "country": "country",
                    "key": "key",
                    "region": "region",
                  },
                  "onLocationSelected": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "currentHourlyForecast": undefined,
                  "isLoading": false,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "currentHourlyForecast": undefined,
                  "isLoading": false,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "hourlyForecasts": Array [],
                  "isLoading": false,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "isLoading": false,
                  "weatherForecasts": Array [],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
